documentation/docs/extensibility.md

```diff
@@ -1,86 +1,214 @@
 
-If you have multiple projects where you want to use a custom template, you could implement this similarly to [piperPipeline](https://github.com/SAP/jenkins-library/blob/master/vars/piperPipeline.groovy).
+The following screenshot shows an example of the configuration in Jenkins.
+Note that the name (1) must be the same as the one you use in your `Jenkinsfile`.
 
-!!! note "How to not get decoupled"
-    Typically, providing a custom template decouples you from centrally provided updates to your template including the stages.<br />
-    Where applicable, you can re-use the stage implementations. This means, you will call e.g. `piperPipelineStageBuild()` as you can see in [piperPipeline](https://github.com/SAP/jenkins-library/blob/master/vars/piperPipeline.groovy).
+![Library Setup](images/customPipelineLib.png "Library Setup")
 
-    Using this approach you can at least benefit from innovations done in individual project "Piper" stages.
+The `Jenkinsfile` of your individual projects would look similar to the following:
 
-!!! note "When to go with a custom template"
-    If the configuration possibilities are not sufficient for you and if _1. Stage Exits_ is not applicable.
+```groovy
+@Library(['piper-lib-os@vINSERT_VERSION_HERE','my-own-pipeline@vINSERT_VERSION_HERE']) _
 
-## 3. Custom Jenkinsfile
+myCustomPipeline script: this
+```
 
-Since project "Piper" fully builds on [Jenkins Pipelines as Code](https://jenkins.io/doc/book/pipeline-as-code/), you can also go with your complete custom `Jenkinsfile`.
+Be sure to adapt the names and version identifiers accordingly, as described in _How to stay up-to-date_.
 
-!!! warning "Decoupling"
-    If you go this route you will be decoupled from the innovations provided with project "Piper", unless you re-use for example stages (as indicated above under _2. Central Custom Templates_).
+### How to stay up-to-date
 
-    **We recommend to use this only as last option for extensibility.**
+Regardless of which of the above options you choose, one downside of this approach is that your pipeline will be out of sync with the centrally maintained pipelines at some point in time.
+We strongly recommend doing _as little modification as possible_ to fulfil your requirements.
+Please be aware that stages may have dependencies on each other.
 
+!!! warning "Don't depend on stage implementation details"
+    Your pipeline should treat _stages_ as a black box, the stage implementations are not a published API and may be subject to change at any time.
 
-## Further tips and information
+!!! warning "Beware of breaking changes"
+    Please be aware that when using the `master` branch of a library, it might always happen that breaking changes occur.
+    We recommend to always fix versions to a released version like in this example: `@Library('my-shared-library@v1.0') _`<br />
```

Comment:
For the Open Source library I believe that we once agreed to recommend to always use `master`.

In the end it should be a decision of the pipeline itself.
For the general purpose one we use `master`:
https://sap.github.io/jenkins-library/stages/introduction/#2-create-jenkinsfile

For the section "How to stay up-to date" we may go with a more fine-grain recommendation. 
If you have a custom pipeline and create yourself certain dependencies you may have a fixed version. 
If you just re-use stages as a black-box you may well go with `master`

Generally, we should be cautious with recommendations in order to not be inconsistent.